### PR TITLE
update support for torch dynamo compile

### PR DESCRIPTION
**Pull request by pacman100**
### What does this PR do?
1. update support for torch dynamo compile

### CODE REVIEW

File: src/accelerate/accelerator.py [hunk 1]
```diff
@@ -1162,10 +1162,10 @@ def prepare_model(self, model: torch.nn.Module, device_placement=None):
         if self.distributed_type == DistributedType.TPU and self.state.fork_launched:
             model = xmp.MpModelWrapper(model).to(self.device)
         # torch.compile should be called last.
-        if self.state.dynamo_backend != DynamoBackend.NO:
-            import torch._dynamo as dynamo
-
-            model = dynamo.optimize(self.state.dynamo_backend.value.lower())(model)
+        if self.state.dynamo_plugin.backend != DynamoBackend.NO:
+            if not hasattr(torch, "compile"):
+                raise ValueError("Using torch.compile requires a nightly install of PyTorch.")
```
Comment (sgugger) [inline src/accelerate/accelerator.py:1167]:
```suggestion
                raise ValueError("Using torch.compile requires PyTorch 2.0 or higher.")
```
PyTorch 2.0 will be out very soon so let's adapt the message here.

File: src/accelerate/commands/config/cluster.py [hunk 2]
```diff
@@ -100,32 +101,47 @@ def get_cluster_input():
...
-                "ipex",
-            ],
+            [x.lower() for x in DYNAMO_BACKENDS],
             _convert_dynamo_backend,
             default=2,
         )
-    else:
-        dynamo_backend = DynamoBackend.NO
+        use_custom_options = _ask_field(
+            "Do you want to customize the defaults? [yes/NO]: ",
```
Comment (sgugger) [inline src/accelerate/commands/config/cluster.py:120]:
```suggestion
            "Do you want to customize the defaults sent to torch.compile? [yes/NO]: ",
```

File: src/accelerate/commands/config/sagemaker.py [hunk 3]
```diff
@@ -170,32 +171,47 @@ def get_sagemaker_input():
...
-                "ipex",
-            ],
+            [x.lower() for x in DYNAMO_BACKENDS],
             _convert_dynamo_backend,
             default=2,
         )
-    else:
-        dynamo_backend = DynamoBackend.NO
+        use_custom_options = _ask_field(
+            "Do you want to customize the defaults? [yes/NO]: ",
```
Comment (sgugger) [inline src/accelerate/commands/config/sagemaker.py:190]:
```suggestion
            "Do you want to customize the defaults sent to torch.compile? [yes/NO]: ",
```

File: src/accelerate/utils/dataclasses.py [hunk 4]
```diff
@@ -347,6 +347,45 @@ def __post_init__(self):
...
+class TorchDynamoPlugin(KwargsHandler):
+    """
+    This plugin is used to compile a model with PyTorch 2.0
+    """
+
+    backend: DynamoBackend = field(
+        default=None,
+        metadata={
+            "help": "Possible options are 'no', 'eager', 'aot_eager', 'inductor', 'nvfuser', 'aot_nvfuser', "
+            "'aot_cudagraphs', 'ofi', 'fx2trt', 'onnxrt', 'ipex'"
```
Comment (sgugger) [inline src/accelerate/utils/dataclasses.py:360]: Maybe use the constant here to make sure the choices are up to do date?

**Review by sgugger — APPROVED** [review]
Thanks a lot for working on this!

File: src/accelerate/utils/dataclasses.py (hunk 4 above)
Comment (pacman100) [reply src/accelerate/utils/dataclasses.py]: Done.

**Pull request merged**